File: src/pyWATS/mes/production.py
"""
MES Production Module

Handles unit information, production operations, and unit lifecycle management.
This module mirrors the Interface.MES Production functionality.
"""

from typing import Optional, List, Union, Dict, Any
from datetime import datetime

from .base import MESBase
from .models import (
    UnitInfo, UnitHistory, UnitVerificationResponse, UnitPhase, 
    IdentifyUnitRequest, StatusEnum, MESResponse
)
from ..rest_api.client import WATSClient
from ..connection import WATSConnection


class Production(MESBase):
    """
    Production management for WATS MES.
    
    Provides functionality for:
    - Unit information retrieval and management
    - Unit lifecycle operations (create, update, phase changes)
    - Parent/child unit relationships
    - Unit history and verification
    """
    
    def __init__(self, connection: Optional[Union[WATSConnection, WATSClient]] = None):
        """
        Initialize Production module.
        
        Args:
            connection: WATS connection or client instance
        """
        super().__init__(connection)
    
    def is_connected(self) -> bool:
        """
        Check if connected to WATS MES Server.
        
        Returns:
            True if connected, False otherwise
        """
        try:
            response = self._client.get("/api/internal/Production/isConnected")
            return response.status_code == 200
        except Exception:
            return False
    
    def get_unit_info(
        self, 
        serial_number: str, 
        part_number: Optional[str] = None
    ) -> Optional[UnitInfo]:
        """
        Get unit information without GUI.
        
        Args:
            serial_number: Unit serial number
            part_number: Optional part number for validation
            
        Returns:
            UnitInfo object or None if not found
            
        Raises:
            WATSAPIException: On API errors
        """
        params = {"serialNumber": serial_number}
        if part_number:
            params["partNumber"] = part_number
        
        try:
            # ?? INTERNAL API: Uses internal production endpoint
            response_data = self._rest_get_json(
                "/api/internal/Production/GetUnitInfo", 
                response_type=UnitInfo
            )
            if isinstance(response_data, UnitInfo):
                return response_data
            elif isinstance(response_data, dict):
                return UnitInfo.model_validate(response_data)  # type: ignore
            return None
        except Exception:
            return None
    
    def identify_uut(
        self,
        part_number: Optional[str] = None,
        serial_number: Optional[str] = None,
        include_test_operation: bool = False,
        select_test_operation: bool = True,
        custom_text: Optional[str] = None,
        always_on_top: bool = True,
        use_workflow: bool = False,
        workflow_status: StatusEnum = StatusEnum.RELEASED,
        context: Optional[Dict[str, Any]] = None
    ) -> Optional[UnitInfo]:
        """
        Display unit identification dialog.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            part_number: Optional part number filter
            serial_number: Optional serial number pre-fill
            include_test_operation: Include test operation in dialog
            select_test_operation: Allow test operation selection
            custom_text: Custom dialog text
            always_on_top: Keep dialog on top
            use_workflow: Use workflow context
            workflow_status: Workflow status filter
            context: Additional context data
            
        Returns:
            Selected UnitInfo or None if cancelled
            
        Raises:
            WATSAPIException: On API errors
        """
        request = IdentifyUnitRequest(
            partNumber=part_number,
            serialNumber=serial_number,
            includeTestOperation=include_test_operation,
            selectTestOperation=select_test_operation,
            customText=custom_text,
            alwaysOnTop=always_on_top,
            useWorkflow=use_workflow,
            workflowStatus=workflow_status,
            context=context or {}
        )
        
        response_data = self._rest_post_json(
            "/api/internal/Production/IdentifyUUT",
            request,
            response_type=UnitInfo
        )
        if isinstance(response_data, UnitInfo):
            return response_data
        elif isinstance(response_data, dict):
            return UnitInfo.model_validate(response_data)  # type: ignore
        return None
    
    def set_unit_process(
        self, 
        serial_number: str, 
        part_number: str, 
        process_name: str
    ) -> bool:
        """
        Set unit's process.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            serial_number: Unit serial number
            part_number: Unit part number
            process_name: Process name to set
            
        Returns:
            True if successful
            
        Raises:
            WATSAPIException: On API errors
        """
        data = {
            "serialNumber": serial_number,
            "partNumber": part_number,
            "processName": process_name
        }
        
        response = self._rest_post_json(
            "/api/internal/Production/SetUnitProcess",
            data,
            response_type=MESResponse
        )
        if isinstance(response, MESResponse):
            return response.success
        elif isinstance(response, dict):
            return response.get("success", False)
        return False
    
    def set_unit_phase(
        self, 
        serial_number: str, 
        part_number: str, 
        phase: Union[UnitPhase, str]
    ) -> bool:
        """
        Set unit's phase.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            serial_number: Unit serial number
            part_number: Unit part number
            phase: Phase to set (UnitPhase enum or string)
            
        Returns:
            True if successful
            
        Raises:
            WATSAPIException: On API errors
        """
        phase_value = phase.value if isinstance(phase, UnitPhase) else phase
        
        data = {
            "serialNumber": serial_number,
            "partNumber": part_number,
            "phase": phase_value
        }
        
        response = self._rest_post_json(
            "/api/internal/Production/SetUnitPhase",
            data,
            response_type=MESResponse
        )
        if isinstance(response, MESResponse):
            return response.success
        elif isinstance(response, dict):
            return response.get("success", False)
        return False
    
    def get_unit_process(self, serial_number: str, part_number: str) -> Optional[str]:
        """
        Get unit's current process.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            serial_number: Unit serial number
            part_number: Unit part number
            
        Returns:
            Process name or None if not found
            
        Raises:
            WATSAPIException: On API errors
        """
        params = {
            "serialNumber": serial_number,
            "partNumber": part_number
        }
        
        try:
            response = self._rest_get_json("/api/internal/Production/GetUnitProcess")
            return response.get("processName")
        except Exception:
            return None
    
    def get_unit_phase(
        self, 
        serial_number: str, 
        part_number: str
    ) -> Optional[UnitPhase]:
        """
        Get unit's current phase.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            serial_number: Unit serial number
            part_number: Unit part number
            
        Returns:
            UnitPhase or None if not found
            
        Raises:
            WATSAPIException: On API errors
        """
        params = {
            "serialNumber": serial_number,
            "partNumber": part_number
        }
        
        try:
            response = self._rest_get_json("/api/internal/Production/GetUnitPhase")
            phase_value = response.get("phase")
            return UnitPhase(phase_value) if phase_value is not None else None
        except Exception:
            return None
    
    def get_unit_phase_string(
        self, 
        serial_number: str, 
        part_number: str
    ) -> Optional[str]:
        """
        Get unit's phase as string.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            serial_number: Unit serial number
            part_number: Unit part number
            
        Returns:
            Phase name or None if not found
            
        Raises:
            WATSAPIException: On API errors
        """
        phase = self.get_unit_phase(serial_number, part_number)
        return phase.name if phase else None
    
    def get_unit_state_history(
        self, 
        serial_number: str, 
        part_number: str
    ) -> int:
        """
        Get unit state/phase change history count.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            serial_number: Unit serial number
            part_number: Unit part number
            
        Returns:
            Number of state changes
            
        Raises:
            WATSAPIException: On API errors
        """
        params = {
            "serialNumber": serial_number,
            "partNumber": part_number
        }
        
        response = self._rest_get_json("/api/internal/Production/GetUnitStateHistory")
        return response.get("count", 0)
    
    def get_unit_history(
        self,
        serial_number: str,
        part_number: Optional[str] = None,
        details: bool = False
    ) -> List[UnitHistory]:
        """
        Get unit history list.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            serial_number: Unit serial number
            part_number: Optional unit part number
            details: Include detailed information
            
        Returns:
            List of UnitHistory objects
            
        Raises:
            WATSAPIException: On API errors
        """
        params = {
            "serialNumber": serial_number,
            "details": details
        }
        if part_number:
            params["partNumber"] = part_number
        
        response = self._rest_get_json("/api/internal/Production/GetUnitHistory")
        history_data = response.get("history", [])
        
        return [UnitHistory.parse_obj(item) for item in history_data]
    
    def set_parent(
        self, 
        serial_number: str, 
        parent_serial_number: str
    ) -> bool:
        """
        Create parent/child unit relationship.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            serial_number: Child unit serial number
            parent_serial_number: Parent unit serial number
            
        Returns:
            True if successful
            
        Raises:
            WATSAPIException: On API errors
        """
        data = {
            "serialNumber": serial_number,
            "parentSerialNumber": parent_serial_number
        }
        
        response = self._rest_post_json(
            "/api/internal/Production/SetParent",
            data,
            response_type=MESResponse
        )
        if isinstance(response, MESResponse):
            return response.success
        elif isinstance(response, dict):
            return response.get("success", False)
        return False
    
    def create_unit(
        self,
        serial_number: str,
        part_number: str,
        revision: str,
        batch_number: str
    ) -> bool:
        """
        Create a new unit.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            serial_number: Unit serial number
            part_number: Unit part number
            revision: Unit revision
            batch_number: Batch number
            
        Returns:
            True if successful
            
        Raises:
            WATSAPIException: On API errors
        """
        data = {
            "serialNumber": serial_number,
            "partNumber": part_number,
            "revision": revision,
            "batchNumber": batch_number
        }
        
        response = self._rest_post_json(
            "/api/internal/Production/CreateUnit",
            data,
            response_type=MESResponse
        )
        if isinstance(response, MESResponse):
            return response.success
        elif isinstance(response, dict):
            return response.get("success", False)
        return False
    
    def add_child_unit(
        self,
        culture_code: str,
        parent_serial_number: str,
        parent_part_number: str,
        child_serial_number: str,
        child_part_number: str,
        check_part_number: str,
        check_revision: str
    ) -> bool:
        """
        Add child unit to parent (box build).
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            culture_code: Culture code for localization
            parent_serial_number: Parent unit serial number
            parent_part_number: Parent unit part number
            child_serial_number: Child unit serial number
            child_part_number: Child unit part number
            check_part_number: Part number to validate
            check_revision: Revision to validate
            
        Returns:
            True if successful
            
        Raises:
            WATSAPIException: On API errors
        """
        data = {
            "cultureCode": culture_code,
            "parentSerialNumber": parent_serial_number,
            "parentPartNumber": parent_part_number,
            "childSerialNumber": child_serial_number,
            "childPartNumber": child_part_number,
            "checkPartNumber": check_part_number,
            "checkRevision": check_revision
        }
        
        response = self._rest_post_json(
            "/api/internal/Production/AddChildUnit",
            data,
            response_type=MESResponse
        )
        if isinstance(response, MESResponse):
            return response.success
        elif isinstance(response, dict):
            return response.get("success", False)
        return False
    
    def remove_child_unit(
        self,
        culture_code: str,
        parent_serial_number: str,
        parent_part_number: str,
        child_serial_number: str,
        child_part_number: str
    ) -> bool:
        """
        Remove child unit from parent.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            culture_code: Culture code for localization
            parent_serial_number: Parent unit serial number
            parent_part_number: Parent unit part number
            child_serial_number: Child unit serial number
            child_part_number: Child unit part number
            
        Returns:
            True if successful
            
        Raises:
            WATSAPIException: On API errors
        """
        data = {
            "cultureCode": culture_code,
            "parentSerialNumber": parent_serial_number,
            "parentPartNumber": parent_part_number,
            "childSerialNumber": child_serial_number,
            "childPartNumber": child_part_number
        }
        
        response = self._rest_post_json(
            "/api/internal/Production/RemoveChildUnit",
            data,
            response_type=MESResponse
        )
        if isinstance(response, MESResponse):
            return response.success
        elif isinstance(response, dict):
            return response.get("success", False)
        return False
    
    def remove_all_child_units(
        self,
        culture_code: str,
        parent_serial_number: str,
        parent_part_number: str
    ) -> bool:
        """
        Remove all child units from parent.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            culture_code: Culture code for localization
            parent_serial_number: Parent unit serial number
            parent_part_number: Parent unit part number
            
        Returns:
            True if successful
            
        Raises:
            WATSAPIException: On API errors
        """
        data = {
            "cultureCode": culture_code,
            "parentSerialNumber": parent_serial_number,
            "parentPartNumber": parent_part_number
        }
        
        response = self._rest_post_json(
            "/api/internal/Production/RemoveAllChildUnits",
            data,
            response_type=MESResponse
        )
        if isinstance(response, MESResponse):
            return response.success
        elif isinstance(response, dict):
            return response.get("success", False)
        return False
    
    def update_unit(
        self,
        serial_number: str,
        part_number: str,
        new_part_number: str,
        new_revision: str
    ) -> bool:
        """
        Update unit's part number and revision.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            serial_number: Unit serial number
            part_number: Current part number
            new_part_number: New part number
            new_revision: New revision
            
        Returns:
            True if successful
            
        Raises:
            WATSAPIException: On API errors
        """
        data = {
            "serialNumber": serial_number,
            "partNumber": part_number,
            "newPartNumber": new_part_number,
            "newRevision": new_revision
        }
        
        response = self._rest_post_json(
            "/api/internal/Production/UpdateUnit",
            data,
            response_type=MESResponse
        )
        if isinstance(response, MESResponse):
            return response.success
        elif isinstance(response, dict):
            return response.get("success", False)
        return False
    
    def update_unit_tag(
        self,
        serial_number: str,
        part_number: str,
        tag_name: str,
        tag_value: str
    ) -> bool:
        """
        Add or update unit tag value.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            serial_number: Unit serial number
            part_number: Unit part number
            tag_name: Tag name
            tag_value: Tag value
            
        Returns:
            True if successful
            
        Raises:
            WATSAPIException: On API errors
        """
        data = {
            "serialNumber": serial_number,
            "partNumber": part_number,
            "tagName": tag_name,
            "tagValue": tag_value
        }
        
        response = self._rest_post_json(
            "/api/internal/Production/UpdateUnitTag",
            data,
            response_type=MESResponse
        )
        if isinstance(response, MESResponse):
            return response.success
        elif isinstance(response, dict):
            return response.get("success", False)
        return False
    
    def get_unit_verification(
        self,
        serial_number: str,
        part_number: Optional[str] = None
    ) -> Optional[UnitVerificationResponse]:
        """
        Get unit verification response.
        
        ?? INTERNAL API: This method uses internal WATS endpoints
        
        Args:
            serial_number: Unit serial number
            part_number: Optional unit part number
            
        Returns:
            UnitVerificationResponse or None if not found
            
        Raises:
            WATSAPIException: On API errors
        """
        params = {"serialNumber": serial_number}
        if part_number:
            params["partNumber"] = part_number
        
        try:
            response = self._rest_get_json(
                "/api/internal/Production/GetUnitVerification",
                response_type=UnitVerificationResponse
            )
            if isinstance(response, UnitVerificationResponse):
                return response
            elif isinstance(response, dict):
                return UnitVerificationResponse.model_validate(response)  # type: ignore
            return None
        except Exception:
            return None

File: referenced_code/Interface.MES/Production/Production.cs
extern alias newclientapi;

using napi = newclientapi::Virinco.WATS.Interface.MES;
using napict = newclientapi::Virinco.WATS.Service.MES.Contract;
using ct = Virinco.WATS.Service.MES.Contract;
using System.Collections.Generic;
using System;
using System.Linq;

namespace Virinco.WATS.Interface.MES.Production
{
    /// <summary>
    /// Class provided to handle unit info (on serialnumber/parnumber)
    /// </summary>
    public class Production : MesBase
    {
        private napi.Production.Production _instance;

        internal Production(napi.Production.Production production)
        {
            this._instance = production;
        }


        public new bool isConnected()
            => _instance.isConnected();

        /// <summary>
        /// If <see cref="IdentifyUUT(out bool, string)"/> is called and this is set to true,
        /// the Display TreeView checkbox is checked. If the checkbox
        /// is not checked, the form will automatically close when the serialnumber
        /// ends with \r\n
        /// </summary>
        public bool DisplayTreeView
        {
            get => _instance.DisplayTreeView;
            set => _instance.DisplayTreeView = value;
        }

        /// <summary>
        /// Holds information about the last scanned serialnumber
        /// </summary>
        public string LastScannedSerialnumber
        {
            get => _instance.LastScannedSerialnumber;
            set => _instance.LastScannedSerialnumber = value;
        }

        /// <summary>
        /// Return <see cref="UnitInfo"/> object (no GUI)
        /// </summary>
        /// <param name="SerialNumber"></param>
        /// <param name="PartNumber"></param>
        /// <returns>A <see cref="UnitInfo"/> object with info about the SN supplied.</returns>
        public UnitInfo GetUnitInfo(string SerialNumber, string PartNumber = "")
            => new UnitInfo(_instance.GetUnitInfo(SerialNumber, PartNumber));

        /// <summary>
        /// Displays SN textbox if connected to service
        /// </summary>
        /// <returns>A <see cref="UnitInfo"/> object of the scanned unit.</returns>
        /// <param name="Continue">Boolean value indicating if the process was stopped</param>        
        /// <param name="PartNumber">PartNumber of unit to identify (used in case of none unique serialnumbers)</param>
        public UnitInfo IdentifyUUT(out bool Continue, string PartNumber = "")
            => new UnitInfo(_instance.IdentifyUUT(out Continue, PartNumber));

        /// <summary>
        /// Display dialog to prompt for serial number
        /// </summary>
        /// <param name="Continue"></param>
        /// <param name="SelectedTestOperation"></param>
        /// <param name="SerialNumber"></param>
        /// <param name="PartNumber"></param>
        /// <param name="IncludeTestOperation"></param>
        /// <param name="SelectTestOperation"></param>
        /// <param name="CustomText"></param>
        /// <param name="AlwaysOnTop"></param>
        /// <param name="UseWorkflow"></param>
        /// <param name="WorkflowStatus"></param>
        /// <param name="context"></param>
        /// <returns></returns>
        public UnitInfo IdentifyUUT(out bool Continue, ref Virinco.WATS.Service.MES.Contract.Process SelectedTestOperation, string SerialNumber = "", string PartNumber = "", bool IncludeTestOperation = false, bool SelectTestOperation = true, string CustomText = null, bool AlwaysOnTop = true, bool UseWorkflow = false, ct.StatusEnum WorkflowStatus = ct.StatusEnum.Released, Dictionary<string, object> context = null)
        {
            napict.Process p = SelectedTestOperation.NewApiInstance;
            var res = _instance.IdentifyUUT(out Continue, ref p, SerialNumber, PartNumber, IncludeTestOperation, SelectTestOperation, CustomText, AlwaysOnTop, UseWorkflow, (napict.StatusEnum)(int)WorkflowStatus, context);
            SelectedTestOperation.NewApiInstance = p;
            return new UnitInfo(res);
        }
        //=> _instance.IdentifyUUT(out Continue, ref SelectedTestOperation, SerialNumber, PartNumber, IncludeTestOperation, SelectTestOperation, CustomText, AlwaysOnTop, UseWorkflow, WorkflowStatus, context);

        /// <summary>
        /// Sets a unit's process to a given value
        /// </summary>
        /// <param name="SerialNumber">Unit's serial number</param>
        /// <param name="PartNumber">Unit's part number</param>
        /// <param name="ProcessName">Unit's new processname</param>        
        public void SetUnitProcess(string SerialNumber, string PartNumber, string ProcessName)
            => _instance.SetUnitProcess(SerialNumber, PartNumber, ProcessName);

        /// <summary>
        /// Sets a unit's phase.
        /// </summary>
        /// <param name="SerialNumber">Unit's serial number</param>
        /// <param name="PartNumber">Unit's part number</param>
        /// <param name="Phase">Unit's new phase,</param>
        public void SetUnitPhase(string SerialNumber, string PartNumber, ct.Unit_Phase Phase)
            => _instance.SetUnitPhase(SerialNumber, PartNumber, (napict.Unit_Phase)(int)Phase);

        /// <summary>
        /// Sets a unit's phase.
        /// </summary>
        /// <param name="SerialNumber">Unit's serial number</param>
        /// <param name="PartNumber">Unit's part number</param>
        /// <param name="Phase">Unit's new phase,</param>
        public void SetUnitPhase(string SerialNumber, string PartNumber, string Phase)
            => _instance.SetUnitPhase(SerialNumber, PartNumber, Phase);

        /// <summary>
        /// Get a unit's current process
        /// </summary>
        /// <param name="SerialNumber">Unit's serial number</param>
        /// <param name="PartNumber">Unit's part number</param>
        /// <returns>Unit's current process</returns>
        public string GetUnitProcess(string SerialNumber, string PartNumber)
            => _instance?.GetUnitProcess(SerialNumber, PartNumber);

        /// <summary>
        /// Get a unit's phase.         
        /// </summary>
        /// <param name="SerialNumber">Unit's serial number</param>
        /// <param name="PartNumber">Unit's part number</param>        
        /// <returns>Units current Phase enum</returns>
        public ct.Unit_Phase GetUnitPhase(string SerialNumber, string PartNumber)
            => (ct.Unit_Phase)(int)_instance.GetUnitPhase(SerialNumber, PartNumber);

        /// <summary>
        /// Get a unit's phase.         
        /// </summary>
        /// <param name="SerialNumber">Unit's serial number</param>
        /// <param name="PartNumber">Unit's part number</param>       
        /// <returns>Units current Phase as string</returns>
        public string GetUnitPhaseString(string SerialNumber, string PartNumber)
            => _instance.GetUnitPhaseString(SerialNumber, PartNumber);

        /// <summary>
        /// Returns unit history of state / phase changes. Changes are returned in three corresponding arrays due to index. Latest change is found on index 0;
        /// </summary>
        /// <param name="serialNumber"></param>
        /// <param name="partNumber"></param>
        /// <param name="states">Returned array of previous states</param>
        /// <param name="phases">Returned array of previous phases</param>
        /// <param name="dateTime">Returned array of date / time of change</param>
        /// <returns>Arrays + return value -1 if error, number of changes (index range) if ok</returns>
        public int GetUnitStateHistory(string serialNumber, string partNumber, out string[] states, out string[] phases, out DateTime[] dateTime)
            => _instance.GetUnitStateHistory(serialNumber, partNumber, out states, out phases, out dateTime);

        /// <summary>
        /// Returns list of unit history in order of newest to latest.
        /// </summary>
        /// <param name="serialNumber">Unit serial number</param>
        /// <param name="partNumber">Unit part number</param>
        /// <param name="details">If <c>true</c>, includes details (info and error messages) for the unit.</param>
        public IEnumerable<UnitHistory> GetUnitHistory(string serialNumber, string partNumber = null, bool details = false)
            => _instance.GetUnitHistory(serialNumber, partNumber, details).Select(i => new UnitHistory(i));

        /// <summary>
        /// Create a parent/child relation between two units.
        /// </summary>
        /// <param name="ParentSerialNumber">Serial Number of parent unit.</param>
        /// <param name="SerialNumber">Serial Number of child unit.</param>
        /// <returns>Value indicating if the action was successfull or not.</returns>
        public bool SetParent(string SerialNumber, string ParentSerialNumber)
            => _instance.SetParent(SerialNumber, ParentSerialNumber);

        /// <summary>
        /// Creates a new unit. The part number (product) and revision will be created if they don't exist.
        /// </summary>
        /// <param name="SerialNumber">Unit's serial number.</param>
        /// <param name="PartNumber">Unit's part number.</param>
        /// <param name="Revision">Unit's revision.</param>
        /// <param name="batchNumber">Unit's batch number.</param>
        /// <returns>True if unit was created.</returns>
        public bool CreateUnit(string SerialNumber, string PartNumber, string Revision, string batchNumber)
            => _instance.CreateUnit(SerialNumber, PartNumber, Revision, batchNumber);

        /// <summary>
        /// Adds a child unit. If Neither CheckPartNumber or CheckRevision is given ProductRelations according to product manager are checked
        /// </summary>
        /// <param name="CultureCode">Culture code used to translate the <paramref name="message"/>.</param>
        /// <param name="ParentSerialNumber">Serial number of the parent unit.</param>
        /// <param name="ParentPartNumber">Part number of the parent unit.</param>
        /// <param name="ChildSerialNumber">Serial number of the child unit to add as child.</param>
        /// <param name="ChildPartNumber">Part number of the child unit to add as child.</param>
        /// <param name="CheckPartNumber">If given, will verify that <paramref name="ChildPartNumber"/> is the same as this. If null, will use Box Build in WATS to verify that the child unit is allowed to be a child of the parent unit.</param>
        /// <param name="CheckRevision">If given, will verify that <paramref name="ChildPartNumber"/> is the same as this. If null, will use Box Build in WATS to verify that the child unit is allowed to be a child of the parent unit.</param>
        /// <param name="message">Translated response message from WATS.</param>
        public bool AddChildUnit(string CultureCode, string ParentSerialNumber, string ParentPartNumber, string ChildSerialNumber, string ChildPartNumber, string CheckPartNumber, string CheckRevision, out string message)
            => _instance.AddChildUnit(CultureCode, ParentSerialNumber, ParentPartNumber, ChildSerialNumber, ChildPartNumber, CheckPartNumber, CheckRevision, out message);

        /// <summary>
        /// Removes a child unit of a given unit.
        /// </summary>
        /// <param name="CultureCode">Culture code used to translate the <paramref name="message"/>.</param>
        /// <param name="ParentSerialNumber">Serial number of the parent unit.</param>
        /// <param name="ParentPartNumber">Part number of the parent unit.</param>
        /// <param name="ChildSerialNumber">Serial number of the child unit to remove as child.</param>
        /// <param name="ChildPartNumber">Part number of the child unit to remove as child.</param>
        /// <param name="message">Translated response message from WATS.</param>
        public bool RemoveChildUnit(string CultureCode, string ParentSerialNumber, string ParentPartNumber, string ChildSerialNumber, string ChildPartNumber, out string message)
            => _instance.RemoveChildUnit(CultureCode, ParentSerialNumber, ParentPartNumber, ChildSerialNumber, ChildPartNumber, out message);

        /// <summary>
        /// Removes all children of a given unit.
        /// </summary>
        /// <param name="CultureCode">Culture code used to translate the <paramref name="message"/></param>
        /// <param name="ParentSerialNumber">Serial number of the parent unit to remove child units from.</param>
        /// <param name="ParentPartNumber">Part number of the parent unit to remove child units from.</param>
        /// <param name="message">Translated response message from WATS.</param>
        /// <returns></returns>
        public bool RemoveAllChildUnits(string CultureCode, string ParentSerialNumber, string ParentPartNumber, out string message)
            => _instance.RemoveAllChildUnits(CultureCode, ParentSerialNumber, ParentPartNumber, out message);

        /// <summary>
        /// Change a unit's partNumber and revision.
        /// </summary>
        /// <param name="SerialNumber">Serial number of unit to update.</param>
        /// <param name="NewPartNumber">New/updated part number.</param>
        /// <param name="NewRevision">New/updated revision.</param>
        [Obsolete("Use overload with partnumber!")]
        public bool UpdateUnit(string SerialNumber, string NewPartNumber, string NewRevision)
            => _instance.UpdateUnit(SerialNumber, NewPartNumber, NewRevision);

        /// <summary>
        /// Change a unit's partNumber and revision. <paramref name="partNumber"/> can be empty, but is required if units do not have unique serial number.
        /// </summary>
        /// <param name="serialNumber">Serial number of unit to update.</param>
        /// <param name="partNumber">Part number of unit to update. Can be empty, but is required if units do not have unique serial number.</param>
        /// <param name="newPartNumber">New/updated part number.</param>
        /// <param name="newRevision">New/updated revision.</param>
        public bool UpdateUnit(string serialNumber, string partNumber, string newPartNumber, string newRevision)
            => _instance.UpdateUnit(serialNumber, partNumber, newPartNumber, newRevision);

        /// <summary>
        /// Add or update a tag value on a unit.
        /// </summary>
        /// <param name="SerialNumber">Serial number of unit to update.</param>
        /// <param name="AttributeName">Tag key.</param>
        /// <param name="AttributeValue">Tag value.</param>
        [Obsolete("Use UpdateUnitTag with specified partnumber!")]
        public bool UpdateUnitAttribute(string SerialNumber, string AttributeName, string AttributeValue)
            => _instance.UpdateUnitAttribute(SerialNumber, AttributeName, AttributeValue);

        /// <summary>
        /// Add or update a tag value on a unit. <paramref name="partNumber"/> can be empty, but is required if units do not have unique serial number.
        /// </summary>
        /// <param name="serialNumber">Serial number of unit to update.</param>
        /// <param name="partNumber">Part number of unit to update. Can be empty, but is required if units do not have unique serial number.</param>
        /// <param name="tagName">Tag key.</param>
        /// <param name="tagValue">Tag value.</param>
        public bool UpdateUnitTag(string serialNumber, string partNumber, string tagName, string tagValue)
            => _instance.UpdateUnitTag(serialNumber, partNumber, tagName, tagValue);

        public UnitVerificationResponse GetUnitVerification(string serialNumber, string partNumber = null)
            => new UnitVerificationResponse(_instance.GetUnitVerification(serialNumber, partNumber));
    }
}
